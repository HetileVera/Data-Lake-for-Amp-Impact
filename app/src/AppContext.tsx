import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
