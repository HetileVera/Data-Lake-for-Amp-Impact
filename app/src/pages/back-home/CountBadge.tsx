import React from "react";
import { Badge } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { useMetadataStore } from "models/MetadataStore";

export const CountBadge = observer(() => {
  const { tone } = useColorScheme();
  const { isError, isReady, store } = useMetadataStore();

  if (isError || !isReady) return null;

  return (
    <Badge ml={3} p={1} pl={2} pr={2} bg={tone(100)} fontSize="small" borderRadius="full" color={tone(700)} fontWeight="bold">
      {store.selectedObjects.length}
    </Badge>
  );
});
