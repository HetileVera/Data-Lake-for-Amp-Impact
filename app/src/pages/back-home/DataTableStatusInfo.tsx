import React from "react";
import { Box, Button, Alert, AlertIcon, Progress } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { useMetadataStore } from "models/MetadataStore";

export const DataTableStatusInfo = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const { isReady, isError, store } = useMetadataStore();
  const progress = store.loadingPercentage;

  const handleTryAgain = () => {
    store.load();
  };

  return (
    (isError && (
      <Alert status="error" variant="left-accent" mt={6} mb={4} color="red.700" alignItems="flex-end">
        <AlertIcon alignSelf="flex-start" />
        <Box>
          <Box>
            Something went wrong and we are unable get the data schema. This might be an intermittent problem. Wait for a few minutes and try again.
          </Box>
          <Box textAlign="right" w="full" mt={4}>
            <Button colorScheme="red" size="sm" onClick={handleTryAgain} loadingText="Processing" isLoading={!isError}>
              Try Again
            </Button>
          </Box>
        </Box>
      </Alert>
    )) ||
    (isReady && null) || (
      <Box mt={6} mb={4} p={0}>
        <Box textAlign="center" fontWeight="bold" color={tone(600)} fontSize="md" mb={4}>
          Retrieving the data schema, this might take a few seconds
          <br />
          {Math.ceil(progress)} %
        </Box>
        <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={tone(100)} />
      </Box>
    )
  );
});
