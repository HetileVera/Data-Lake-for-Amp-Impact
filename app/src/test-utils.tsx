import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
