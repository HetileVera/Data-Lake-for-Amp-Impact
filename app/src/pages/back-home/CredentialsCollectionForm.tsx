import React from "react";
import { Box, Button } from "@chakra-ui/react";
import { ArrowForwardIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { SubmitHandler, useForm } from "react-hook-form";

import { useInstallation } from "AppContext";
import { useColorScheme } from "models/useColorScheme";
import { CredentialsValidationException } from "api/validate-credentials";

import { CredentialsForm, ICredentialsFormInput } from "pages/step1/CredentialsForm";
import { CredentialsError } from "pages/step1/CredentialsError";

interface ICredentialsCollectionForm {
  onCancel: () => void;
  onDone: () => void;
  buttonTitle?: string;
}

export const CredentialsCollectionForm = observer(({ onCancel, onDone, buttonTitle = "Resume" }: ICredentialsCollectionForm) => {
  const { colorScheme } = useColorScheme();
  const [error, setError] = React.useState<CredentialsValidationException | Error>();
  const installation = useInstallation();

  const onSubmit: SubmitHandler<ICredentialsFormInput> = async (values) => {
    setError(undefined);
    const accessKey = values.accessKeyId;
    const secretKey = values.secretAccessKey;
    const region = installation.region;

    try {
      await installation.connectToAwsStep.connectToAws(accessKey, secretKey, region);
      if (installation.isPostDeployment) installation.metadataStore.reset();

      return onDone();
    } catch (err: any) {
      console.log(err);
      setError(err);
    }
  };

  const {
    handleSubmit,
    register,
    control,
    formState: { errors, isSubmitting },
  } = useForm<ICredentialsFormInput>();

  const buttonProps: any = {};
  if (buttonTitle === "Resume") buttonProps.rightIcon = <ArrowForwardIcon />;

  return (
    <>
      <CredentialsError exception={error} />

      <Box pt={6}>
        {/* @ts-ignore */}
        <CredentialsForm {...{ register, errors, isSubmitting, control }} />
      </Box>

      <Box textAlign="right" w="full" mt={6} mb={0}>
        <Button colorScheme={colorScheme} size="md" disabled={isSubmitting} variant="ghost" onClick={() => onCancel()}>
          Cancel
        </Button>
        <Button
          colorScheme={colorScheme}
          size="md"
          loadingText="Connecting"
          ml={3}
          isLoading={isSubmitting}
          disabled={isSubmitting}
          onClick={handleSubmit(onSubmit)}
          {...buttonProps}
        >
          {buttonTitle}
        </Button>
      </Box>
    </>
  );
});
