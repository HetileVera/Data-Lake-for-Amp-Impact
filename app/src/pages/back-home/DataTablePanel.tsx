import React from "react";
import { Box, Heading } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { useMetadataStore } from "models/MetadataStore";

import { DataModelTablePreview } from "pages/step4/data-model-table-preview/DataModelTablePreview";

import { CountBadge } from "./CountBadge";
import { DataTableStatusInfo } from "./DataTableStatusInfo";

export const DataTablePanel = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const { store, isReady } = useMetadataStore();
  const selectedObjectsCount = store.selectedObjects.length;
  const isEmptyModel = selectedObjectsCount === 0;

  return (
    <>
      <Heading size="md" pt="0px" pb="0px" color={tone(600)} letterSpacing="-1px">
        Objects <CountBadge />
      </Heading>
      {isEmptyModel && isReady && (
        <Box bg={tone(100)} textAlign="center" p={4} mt={6} mb={4} fontSize="sm" color={tone(600)} borderRadius="md">
          No objects
        </Box>
      )}
      <DataTableStatusInfo />
      <DataModelTablePreview colorScheme={colorScheme} showCaption={false} />
    </>
  );
});
