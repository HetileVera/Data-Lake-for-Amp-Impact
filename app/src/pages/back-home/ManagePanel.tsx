import React from "react";
import { Box, Heading, Text, Button } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useStore } from "AppContext";
import { useColorScheme } from "models/useColorScheme";

import { StartNewWarning } from "./StartNewWarning";
import { CredentialsCollectionForm } from "./CredentialsCollectionForm";

export const ManagePanel = observer(() => {
  const store = useStore();
  const { tone, colorScheme } = useColorScheme();
  const [collectCredentials, setCollectCredentials] = React.useState(false);
  const [warning, setWarning] = React.useState(false);

  React.useEffect(() => {
    if (collectCredentials || warning) setTimeout(() => window.scrollBy(0, 300), 200);
  }, [collectCredentials, warning]);

  return (
    <>
      <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6} color={tone(800)}>
        <Heading size="md" pt="0px" pb="10px" color={tone(600)} letterSpacing="-1px">
          Manage data lake
        </Heading>
        <Text>
          You can manage users, see the latest data import status and more. Click <b>Manage</b> to unlock these capabilities. We will ask you for your
          AWS admin credentials so that we can connect to your AWS account.
        </Text>
        {!collectCredentials && !warning && (
          <Box textAlign="right" w="full" mt={6} mb={0}>
            <Button colorScheme={colorScheme} size="md" variant="outline" mr={6} onClick={() => setWarning(true)}>
              Start New
            </Button>
            <Button colorScheme={colorScheme} size="md" onClick={() => setCollectCredentials(true)}>
              Manage
            </Button>
          </Box>
        )}
      </Box>

      {warning && <StartNewWarning onCancel={() => setWarning(false)} onContinue={() => store.newInstallation()} />}

      {collectCredentials && (
        <CredentialsCollectionForm
          buttonTitle="Manage"
          onCancel={() => setCollectCredentials(false)}
          onDone={() => {
            setCollectCredentials(false);
          }}
        />
      )}
    </>
  );
});
