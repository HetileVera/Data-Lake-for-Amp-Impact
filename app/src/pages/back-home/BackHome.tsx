import React from "react";
import { Box, Text, Button } from "@chakra-ui/react";
import { ArrowForwardIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { useNavigate } from "react-router-dom";

import { useInstallation, useStore } from "AppContext";
import { Layout } from "components/Layout";
import { TimeAgo } from "components/TimeAgo";
import { isCompleted } from "models/Installation";
import { useColorScheme } from "models/useColorScheme";
import { AccessInformation } from "pages/step6/AccessInformation";
import { UsersPanel } from "pages/step6/users-panel/UsersPanel";
import { CloudwatchDashboardPanel } from "pages/step6/CloudwatchDashboardPanel";
import { SubscribeSNSPanel } from "pages/step6/SubscribeSNSPanel";
import { DataImportStatusPanel } from "pages/step5/data-import-status-panel/DataImportStatusPanel";

import { ReviewGrid } from "./ReviewGrid";
import { DataTablePanel } from "./DataTablePanel";
import { ManagePanel } from "./ManagePanel";
import { StartNewWarning } from "./StartNewWarning";
import { CredentialsCollectionForm } from "./CredentialsCollectionForm";

export const BackHome = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const navigate = useNavigate();
  const store = useStore();
  const installation = useInstallation();
  const [collectCredentials, setCollectCredentials] = React.useState(false);
  const [warning, setWarning] = React.useState(false);
  const completedProvisioning = isCompleted(installation.deploymentStep);
  const reviewCompleted = isCompleted(installation.reviewStep);
  const credentialsAvailable = !installation.credentials.empty;

  const gotoNextStep = () => {
    const nextStep = installation.nextStepNumber;
    navigate(`/steps/${nextStep}`);
  };

  const handleCollectCredentials = () => {
    if (credentialsAvailable) return gotoNextStep();
    setCollectCredentials(true);
  };

  const handleStartNew = () => {
    if (reviewCompleted) {
      setWarning(true);
    } else {
      store.newInstallation();
    }
  };

  React.useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  React.useEffect(() => {
    if (collectCredentials || warning) setTimeout(() => window.scrollBy(0, 300), 200);
  }, [collectCredentials, warning]);

  return (
    <Layout step={0} title="Data Lake Information">
      <Box position="relative" boxShadow="base" borderRadius="lg" bg={tone(50)} p={10} pb={6} color={tone(800)}>
        <Box fontSize="small" mt={1} mb={8} color={tone(700)}>
          <TimeAgo time={installation.startDate} /> &#8226; by {installation.startedBy}
        </Box>

        <ReviewGrid />

        {!completedProvisioning && !collectCredentials && !warning && (
          <>
            <Text pl={2} mt={6}>
              Click <b>Resume</b> to continue the steps of provisioning the data lake
              {!credentialsAvailable && <>, we will ask you for your AWS admin credentials so that we can connect to your AWS account your </>}. You
              can also choose to start a new data lake.
            </Text>
            <Box textAlign="right" w="full" mt={6} mb={0}>
              <Button colorScheme={colorScheme} size="md" variant="outline" mr={6} onClick={handleStartNew}>
                Start New
              </Button>
              <Button colorScheme={colorScheme} size="md" rightIcon={<ArrowForwardIcon />} onClick={handleCollectCredentials}>
                Resume
              </Button>
            </Box>
          </>
        )}
      </Box>

      {completedProvisioning && (
        <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={10}>
          <AccessInformation />
        </Box>
      )}

      {completedProvisioning && credentialsAvailable && (
        <>
          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <UsersPanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6}>
            <DataImportStatusPanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6}>
            <DataTablePanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <CloudwatchDashboardPanel region={installation.region} id={installation.id} />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <SubscribeSNSPanel />
          </Box>
        </>
      )}

      {completedProvisioning && !credentialsAvailable && <ManagePanel />}

      {warning && <StartNewWarning onCancel={() => setWarning(false)} onContinue={() => store.newInstallation()} />}

      {collectCredentials && (
        <CredentialsCollectionForm
          onCancel={() => setCollectCredentials(false)}
          onDone={() => {
            setCollectCredentials(false);
            gotoNextStep();
          }}
        />
      )}
      <Box h="50px">&nbsp;</Box>
    </Layout>
  );
});
