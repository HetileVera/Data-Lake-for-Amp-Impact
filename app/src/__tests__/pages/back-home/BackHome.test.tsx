import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

import { render, RenderWithChakra } from "test-utils";

import * as appContext from "AppContext";
import { BackHome } from "pages/back-home/BackHome";

const spyScrollTo = jest.fn();

describe("BackHome component", () => {
  beforeEach(() => {
    Object.defineProperty(global.window, "scrollTo", { value: spyScrollTo });
  });

  afterEach(() => {
    spyScrollTo.mockClear();
    cleanup;
  });

  test("snapshot", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        connectToSalesforceStep: "test-step",
        credentials: [],
        deploymentStep: "1",
        reviewStep: "1",
        nextStepNumber: "2",
      };
    });

    const tree = create(
      <RenderWithChakra>
        <appContext.StoreProvider>
          <BrowserRouter>
            <BackHome />
          </BrowserRouter>
        </appContext.StoreProvider>
      </RenderWithChakra>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        connectToSalesforceStep: "test-step",
        credentials: [],
        deploymentStep: "1",
        reviewStep: "1",
        nextStepNumber: "2",
      };
    });

    render(
      <appContext.StoreProvider>
        <BrowserRouter>
          <BackHome />
        </BrowserRouter>
      </appContext.StoreProvider>
    );

    expect(screen.getByText(/aws account id/i)).toBeInTheDocument();
    expect(screen.getByText(/test-account-id/i)).toBeInTheDocument();
    expect(screen.getByText(/aws region/i)).toBeInTheDocument();
    expect(screen.getByText(/connection name/i)).toBeInTheDocument();
    expect(screen.getByText(/test-appflow-connection-name/i)).toBeInTheDocument();
  });
});
