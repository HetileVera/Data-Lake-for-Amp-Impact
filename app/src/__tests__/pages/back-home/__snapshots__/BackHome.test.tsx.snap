// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackHome component snapshot 1`] = `
Array [
  <div
    className="css-1v55i00"
    data-testid="curved-box"
  />,
  <div
    className="chakra-container css-1szj4qk"
  >
    <div
      className="chakra-container css-1szj4qk"
    >
      <div
        className="chakra-stack css-olw0lh"
      >
        <div
          className="chakra-stack css-18lyn6q"
        >
          <div
            className="css-1i3fafl"
          >
            <img
              className="chakra-image css-0"
              src="Salesforce_Corporate_Logo_RGB.png"
            />
          </div>
          <div
            className="chakra-stack__divider css-r2vkku"
          />
          <div
            className="css-nutmff"
          >
            <img
              className="chakra-image css-ir5tdx"
              src="AWS_logo_RGB_BLK.png"
            />
          </div>
        </div>
        <div
          className="chakra-stack css-1n7gsjr"
        >
          <div
            className="css-18n4g2t"
          >
            <h2
              className="chakra-heading css-1opsxio"
              id="h2-title"
            >
              Data Lake for Nonprofits, Powered by AWS
            </h2>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="chakra-container css-ms16h5"
  >
    <div
      className="css-1w1wwoy"
    >
      <div
        className="css-1dqvb5u"
      >
        <div
          className="css-1rr4qq7"
        >
          <h2
            className="chakra-heading css-1t51nrq"
          >
            Data Lake
          </h2>
          <div
            className="css-133irwf"
          >
            <time
              dateTime="2022-01-01T00:00:00.000Z"
              title="2022-01-01"
            >
              8 months ago
            </time>
             • by 
            test-user
          </div>
        </div>
        <div
          className="css-1t3jdvp"
        >
          <p
            className="chakra-text css-13o7eu2"
          >
            Installation Id
          </p>
          <p
            className="chakra-text css-0"
          >
            1
          </p>
        </div>
      </div>
      <div
        className="css-iektmo"
      >
        <div
          className="css-1912uuf"
        >
          AWS Account Id
        </div>
        <div
          className="css-im2a0n"
        >
          test-account-id
        </div>
        <div
          className="css-1912uuf"
        >
          AWS Region
        </div>
        <div
          className="css-im2a0n"
        >
          US East (N. Virginia)
        </div>
        <div
          className="css-1912uuf"
        >
          Connection Name
        </div>
        <div
          className="css-im2a0n"
        >
          test-appflow-connection-name
        </div>
      </div>
      <p
        className="chakra-text css-1qo0xie"
      >
        Click 
        <b>
          Resume
        </b>
         to continue the steps of provisioning the data lake
        . You can also choose to start a new data lake.
      </p>
      <div
        className="css-s6koyw"
      >
        <button
          className="chakra-button css-wd3f61"
          onClick={[Function]}
          type="button"
        >
          Start New
        </button>
        <button
          className="chakra-button css-mwvyhu"
          onClick={[Function]}
          type="button"
        >
          Resume
          <span
            className="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden={true}
              className="chakra-icon css-onkibi"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
  <div
    className="css-16kb41"
  >
     
  </div>,
  <span />,
]
`;
